From 6afb92d060eacb185654922695a337b9d6faaab1 Mon Sep 17 00:00:00 2001
From: Alexandr Savca <alexandr.savca89@gmail.com>
Date: Fri, 22 Nov 2019 20:58:14 +0200
Subject: [PATCH] ath9k: override regdomain hardcoded in EEPROM with custom
 value

Get your country code from linux/drivers/net/wireless/ath/regd.h
and supply as a parameter:

    modprobe ath9k_hw override_eeprom_regdomain=<CODE>
    modprobe ath9k
---
 drivers/net/wireless/ath/ath9k/hw.c | 24 +++++++++++++++++++++++-
 1 file changed, 23 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath9k/hw.c b/drivers/net/wireless/ath/ath9k/hw.c
index 052deffb4c9d..8648755fdf39 100644
--- a/drivers/net/wireless/ath/ath9k/hw.c
+++ b/drivers/net/wireless/ath/ath9k/hw.c
@@ -37,6 +37,15 @@ MODULE_DESCRIPTION("Support for Atheros 802.11n wireless LAN cards.");
 MODULE_SUPPORTED_DEVICE("Atheros 802.11n WLAN cards");
 MODULE_LICENSE("Dual BSD/GPL");
 
+static int modparam_override_eeprom_regdomain = -1;
+module_param_named(
+	override_eeprom_regdomain,
+	modparam_override_eeprom_regdomain,
+	int,
+	0444);
+MODULE_PARM_DESC(override_eeprom_regdomain,
+	"Override regdomain hardcoded in EEPROM with this value (DANGEROUS).");
+
 static void ath9k_hw_set_clockrate(struct ath_hw *ah)
 {
 	struct ath_common *common = ath9k_hw_common(ah);
@@ -2494,7 +2503,20 @@ int ath9k_hw_fill_cap_info(struct ath_hw *ah)
 	u8 ant_div_ctl1, tx_chainmask, rx_chainmask;
 
 	eeval = ah->eep_ops->get_eeprom(ah, EEP_REG_0);
-	regulatory->current_rd = eeval;
+	if (modparam_override_eeprom_regdomain != -1) {
+		dev_err(ah->dev,
+			"DANGER! You're overriding EEPROM-defined regulatory "
+			"domain.\n"
+			"Your card was not certified to operate on the domain "
+			"you choosed.\n"
+			"This might result in a violation of your local "
+			"regulatory rules.\n"
+			"Do not ever do that unless you really know what you "
+			"do.\n");
+		regulatory->current_rd =
+			modparam_override_eeprom_regdomain + 32768;
+	} else
+		regulatory->current_rd = eeval;
 
 	if (ah->opmode != NL80211_IFTYPE_AP &&
 	    ah->hw_version.subvendorid == AR_SUBVENDOR_ID_NEW_A) {
-- 
2.24.0

